ops/model.py

```diff
@@ -98,6 +91,16 @@
 })
 
 
+# Copied from typeshed.
+_KT = typing.TypeVar("_KT")
+_VT_co = typing.TypeVar("_VT_co", covariant=True)
+
+
+class _SupportsKeysAndGetItem(typing.Protocol[_KT, _VT_co]):
+    def keys(self) -> typing.Iterable[_KT]: ...
+    def __getitem__(self, __key: _KT) -> _VT_co: ...
+
+
```

Comment:
I'm not thrilled about adding this here. However, I couldn't find any other way to satisfy the way one of the charms (Jenkins, I think) was using relation data.

A minimal example is:

```python
class A(typing.TypedDict):
    foo: str

r = ops.RelationDataContent(None, None, None)  # type:ignore
a = A(foo="hello")
r.update(a)
```

mypy complains about this:

```
/tmp/reldata.py:9: error: Argument 1 to "update" of "MutableMapping" has incompatible type "A"; expected "SupportsKeysAndGetItem[str, str]"  [arg-type]
/tmp/reldata.py:9: note: Following member(s) of "A" have conflicts:
/tmp/reldata.py:9: note:     Expected:
/tmp/reldata.py:9: note:         def __getitem__(self, str, /) -> str
/tmp/reldata.py:9: note:     Got:
/tmp/reldata.py:9: note:         def __getitem__(self, str, /) -> object
```

The charm error is the same.

It seems like it shouldn't be required, because `RelationDataContent` has an explicit `__getitem__` that does return `str`, and the `MutableMapping` is `[str, str]`. It might be that this is a limitation of mypy? pyright does complain too, though:

```
/tmp/reldata.py
  /tmp/reldata.py:9:1 - error: No overloads for "update" match the provided arguments (reportCallIssue)
  /tmp/reldata.py:9:10 - error: Argument of type "A" cannot be assigned to parameter "__m" of type "Iterable[tuple[str, str]]" in function "update"
    "A" is incompatible with "Iterable[tuple[str, str]]"
      Type parameter "_T_co@Iterable" is covariant, but "str" is not a subtype of "tuple[str, str]"
        "str" is incompatible with "tuple[str, str]" (reportArgumentType)
2 errors, 0 warnings, 0 informations 
```